product/optee-stm32mp1/module/stm32_pwr_regu/src/mod_stm32_pwr_regu.c

```diff
@@ -0,0 +1,268 @@
+         fwk_id_get_element_idx(dev_id), ctx->pwr_id,
+         mode_id == MOD_VOLTD_MODE_ID_ON ? "on" : "off");
+
+    return FWK_SUCCESS;
+}
+
+static int pwr_regu_get_level(fwk_id_t dev_id, int *level_uv)
+{
+    struct stm32_pwr_regu_dev_ctx *ctx = NULL;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    if (!nsec_can_access_pwr_regu(ctx->pwr_id)) {
+        return FWK_E_ACCESS;
+    }
+
+    *level_uv = pwr_regu_level(ctx->pwr_id);
+
+    FWK_LOG_TRACE(MOD_NAME "SCMI voltd %u: get_level PWR#%u = %d",
+         fwk_id_get_element_idx(dev_id), ctx->pwr_id, *level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static int pwr_regu_set_level(fwk_id_t dev_id, int level_uv)
+{
+    struct stm32_pwr_regu_dev_ctx *ctx = NULL;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    if (!nsec_can_access_pwr_regu(ctx->pwr_id)) {
+        return FWK_E_ACCESS;
+    }
+
+    FWK_LOG_TRACE(MOD_NAME "SCMI voltd %u: set_level PWR#%u to %d",
+         fwk_id_get_element_idx(dev_id), ctx->pwr_id, level_uv);
+
+    if (level_uv != pwr_regu_level(ctx->pwr_id)) {
+        return FWK_E_RANGE;
+    }
+
+    return FWK_SUCCESS;
+}
+
+static int pwr_regu_get_info(fwk_id_t dev_id, struct mod_voltd_info *info)
+{
+    struct stm32_pwr_regu_dev_ctx *ctx = NULL;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    if (!nsec_can_access_pwr_regu(ctx->pwr_id)) {
+        return FWK_E_ACCESS;
+    }
+
+    memset(info, 0, sizeof(*info));
+    info->level_range.level_type = MOD_VOLTD_VOLTAGE_LEVEL_DISCRETE;
+    info->level_range.min_uv = pwr_regu_level(ctx->pwr_id);
+    info->level_range.max_uv = info->level_range.min_uv;
+    info->level_range.level_count = 1;
+    info->name = ctx->name;
+
+    FWK_LOG_TRACE(MOD_NAME"SCMI voltd %u: get_info PWR#%u",
+         fwk_id_get_element_idx(dev_id), ctx->pwr_id);
+
+    return FWK_SUCCESS;
+}
+
+static int pwr_regu_level_from_index(fwk_id_t dev_id, unsigned int index,
+                                     int32_t *level_uv)
+{
+    struct stm32_pwr_regu_dev_ctx *ctx = NULL;
+
+    ctx = module_ctx.dev_ctx_table + fwk_id_get_element_idx(dev_id);
+
+    if (!fwk_module_is_valid_element_id(dev_id)) {
+        return FWK_E_PARAM;
+    }
+
+    if (!nsec_can_access_pwr_regu(ctx->pwr_id)) {
+        return FWK_E_ACCESS;
+    }
+
+    if (index > 0) {
+        return FWK_E_RANGE;
+    }
+
+    *level_uv = pwr_regu_level(ctx->pwr_id);
+
+    FWK_LOG_TRACE(MOD_NAME "SCMI voltd %u: get_level_from_index PWR#%u = %"PRId32,
+         fwk_id_get_element_idx(dev_id), ctx->pwr_id, *level_uv);
+
+    return FWK_SUCCESS;
+}
+
+static const struct mod_voltd_drv_api api_stm32_pwr_regu = {
```

Comment:
I would pop the functions in the order they appear here, for ease of finding things in future :)